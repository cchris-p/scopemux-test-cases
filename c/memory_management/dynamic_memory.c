/**
 * Demonstration of C Memory Management
 *
 * This example covers:
 * - Dynamic memory allocation (malloc, calloc, realloc, free)
 * - Memory manipulation functions (memcpy, memmove, memset)
 * - Pointer arithmetic
 * - Memory safety considerations
 * - Custom memory allocation patterns
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

/**
 * Basic dynamic memory allocation with malloc
 */
void demonstrate_malloc() {
    printf("\n=== Basic malloc and free ===\n");
    
    // Allocate memory for 5 integers
    int* numbers = (int*)malloc(5 * sizeof(int));
    
    // Check if allocation was successful
    if (numbers == NULL) {
        printf("Memory allocation failed\n");
        return;
    }
    
    // Initialize the allocated memory
    for (int i = 0; i < 5; i++) {
        numbers[i] = i * 10;
    }
    
    // Use the allocated memory
    printf("Allocated array contents: ");
    for (int i = 0; i < 5; i++) {
        printf("%d ", numbers[i]);
    }
    printf("\n");
    
    // Free the allocated memory
    free(numbers);
    numbers = NULL; // Best practice: set freed pointers to NULL
    
    // Trying to use freed memory would cause undefined behavior
    // numbers[0] = 100; // This would be a bug!
}

/**
 * Using calloc for zero-initialized memory
 */
void demonstrate_calloc() {
    printf("\n=== Using calloc ===\n");
    
    // Allocate and zero-initialize memory for 5 integers
    int* numbers = (int*)calloc(5, sizeof(int));
    
    if (numbers == NULL) {
        printf("Memory allocation failed\n");
        return;
    }
    
    // calloc initializes memory to zero
    printf("calloc initialized array: ");
    for (int i = 0; i < 5; i++) {
        printf("%d ", numbers[i]);
    }
    printf("\n");
    
    // Now assign some values
    for (int i = 0; i < 5; i++) {
        numbers[i] = (i + 1) * 5;
    }
    
    printf("After assignment: ");
    for (int i = 0; i < 5; i++) {
        printf("%d ", numbers[i]);
    }
    printf("\n");
    
    // Free the allocated memory
    free(numbers);
    numbers = NULL;
}

/**
 * Using realloc to resize allocated memory
 */
void demonstrate_realloc() {
    printf("\n=== Using realloc ===\n");
    
    // Start with a small array
    int* numbers = (int*)malloc(3 * sizeof(int));
    
    if (numbers == NULL) {
        printf("Memory allocation failed\n");
        return;
    }
    
    // Initialize the array
    for (int i = 0; i < 3; i++) {
        numbers[i] = i + 1;
    }
    
    printf("Initial array: ");
    for (int i = 0; i < 3; i++) {
        printf("%d ", numbers[i]);
    }
    printf("\n");
    
    // Resize the array to hold 5 integers
    int* new_numbers = (int*)realloc(numbers, 5 * sizeof(int));
    
    if (new_numbers == NULL) {
        printf("Memory reallocation failed\n");
        free(numbers); // Original memory must still be freed if realloc fails
        return;
    }
    
    // The original data is preserved; realloc copied it for us
    numbers = new_numbers; // Update our pointer
    
    // Initialize the new elements
    numbers[3] = 4;
    numbers[4] = 5;
    
    printf("After realloc (expanded): ");
    for (int i = 0; i < 5; i++) {
        printf("%d ", numbers[i]);
    }
    printf("\n");
    
    // Shrink the array
    new_numbers = (int*)realloc(numbers, 2 * sizeof(int));
    
    if (new_numbers == NULL) {
        printf("Memory reallocation failed\n");
        free(numbers);
        return;
    }
    
    numbers = new_numbers;
    
    printf("After realloc (shrunk): ");
    for (int i = 0; i < 2; i++) {
        printf("%d ", numbers[i]);
    }
    printf("\n");
    
    // Free the memory
    free(numbers);
    numbers = NULL;
}

/**
 * Demonstrate pointer arithmetic and memory access
 */
void demonstrate_pointer_arithmetic() {
    printf("\n=== Pointer Arithmetic ===\n");
    
    // Allocate an array of 5 integers
    int* arr = (int*)malloc(5 * sizeof(int));
    
    if (arr == NULL) {
        printf("Memory allocation failed\n");
        return;
    }
    
    // Initialize array
    for (int i = 0; i < 5; i++) {
        arr[i] = i * 10;
    }
    
    // Basic array indexing
    printf("Array indexing: arr[2] = %d\n", arr[2]);
    
    // Equivalent pointer arithmetic
    printf("Pointer arithmetic: *(arr + 2) = %d\n", *(arr + 2));
    
    // Moving the pointer (unsafe, must keep track of original pointer for freeing)
    int* ptr = arr;
    printf("Original pointer position: *ptr = %d\n", *ptr);
    
    // Move pointer to next element
    ptr++;
    printf("After ptr++: *ptr = %d\n", *ptr);
    
    // Move pointer by an offset
    ptr = arr + 3;
    printf("After ptr = arr + 3: *ptr = %d\n", *ptr);
    
    // Difference between pointers
    int* ptr2 = arr + 4;
    ptrdiff_t diff = ptr2 - ptr;
    printf("Pointer difference (ptr2 - ptr): %td elements\n", diff);
    
    // Free the original pointer, not the moved one
    free(arr);
    arr = NULL;
    ptr = NULL; // Not necessary, but good practice
    ptr2 = NULL;
}

/**
 * Demonstrate memory manipulation functions
 */
void demonstrate_memory_functions() {
    printf("\n=== Memory Manipulation Functions ===\n");
    
    // Allocate memory
    char* buffer1 = (char*)malloc(20);
    char* buffer2 = (char*)malloc(20);
    
    if (buffer1 == NULL || buffer2 == NULL) {
        free(buffer1); // Handle partial allocation failure
        free(buffer2);
        printf("Memory allocation failed\n");
        return;
    }
    
    // memset: Fill memory with a specific byte
    memset(buffer1, 'A', 19);
    buffer1[19] = '\0'; // Null-terminate for printing
    printf("After memset: %s\n", buffer1);
    
    // memcpy: Copy non-overlapping memory
    const char* source = "Hello, World!";
    memcpy(buffer2, source, strlen(source) + 1); // +1 for the null terminator
    printf("After memcpy: %s\n", buffer2);
    
    // memmove: Safe for overlapping regions
    char overlap[20] = "abcdefghijklmnopqrs";
    printf("Before memmove: %s\n", overlap);
    
    // Move "defghi" to position starting at "bcdef"
    memmove(overlap + 1, overlap + 3, 6);
    printf("After memmove: %s\n", overlap);
    
    // memcmp: Compare memory regions
    char str1[] = "Hello";
    char str2[] = "Hello";
    char str3[] = "World";
    
    int cmp1 = memcmp(str1, str2, 5);
    int cmp2 = memcmp(str1, str3, 5);
    
    printf("memcmp(str1, str2) = %d (same content)\n", cmp1);
    printf("memcmp(str1, str3) = %d (different content)\n", cmp2);
    
    // Clean up
    free(buffer1);
    free(buffer2);
    buffer1 = NULL;
    buffer2 = NULL;
}

/**
 * Demonstrate a common pitfall: memory leaks
 */
void demonstrate_memory_leaks() {
    printf("\n=== Memory Leaks ===\n");
    
    // Allocate memory
    int* data = (int*)malloc(sizeof(int) * 10);
    
    if (data == NULL) {
        printf("Memory allocation failed\n");
        return;
    }
    
    printf("Memory allocated at %p\n", (void*)data);
    
    // Example of code paths that might cause leaks
    int error_condition = 0; // Change to 1 to simulate error path
    
    if (error_condition) {
        printf("Error occurred, returning early\n");
        return; // This would leak memory if we forgot to free
    }
    
    // Nested allocations make leak tracking harder
    int* more_data = (int*)malloc(sizeof(int) * 5);
    
    if (more_data == NULL) {
        printf("Second allocation failed\n");
        free(data); // Don't forget to free the first allocation
        return;
    }
    
    printf("More memory allocated at %p\n", (void*)more_data);
    
    // Proper cleanup
    free(more_data);
    free(data);
    more_data = NULL;
    data = NULL;
    
    printf("All memory properly freed\n");
}

/**
 * Demonstrate a simple custom memory pool
 */
#define POOL_SIZE 1024
static char memory_pool[POOL_SIZE];
static size_t pool_offset = 0;

void* pool_alloc(size_t size) {
    // Simple bump allocator
    if (pool_offset + size > POOL_SIZE) {
        return NULL; // Out of pool memory
    }
    
    void* ptr = &memory_pool[pool_offset];
    pool_offset += size;
    
    // Ensure alignment (simplified)
    pool_offset = (pool_offset + 7) & ~7;
    
    return ptr;
}

void pool_reset() {
    pool_offset = 0;
}

void demonstrate_memory_pool() {
    printf("\n=== Custom Memory Pool ===\n");
    
    // Reset the pool to start fresh
    pool_reset();
    
    // Allocate some memory from the pool
    int* numbers = (int*)pool_alloc(5 * sizeof(int));
    
    if (numbers == NULL) {
        printf("Pool allocation failed\n");
        return;
    }
    
    // Initialize the memory
    for (int i = 0; i < 5; i++) {
        numbers[i] = i * 100;
    }
    
    printf("Pool allocated array: ");
    for (int i = 0; i < 5; i++) {
        printf("%d ", numbers[i]);
    }
    printf("\n");
    
    // Allocate another block
    char* text = (char*)pool_alloc(20);
    
    if (text == NULL) {
        printf("Pool allocation failed\n");
        return;
    }
    
    strcpy(text, "Hello from pool");
    printf("Pool allocated string: %s\n", text);
    
    printf("Pool usage: %zu bytes\n", pool_offset);
    
    // No need to free individual allocations with this simple pool
    // Just reset the entire pool when done
    pool_reset();
    printf("Pool reset, usage: %zu bytes\n", pool_offset);
    
    // The memory for numbers and text is now available for reuse
    // but accessing them would be dangerous as the memory may be reused
}

/**
 * Demonstrate handling of multi-dimensional arrays with dynamic allocation
 */
void demonstrate_multidimensional_arrays() {
    printf("\n=== Multi-dimensional Arrays ===\n");
    
    int rows = 3;
    int cols = 4;
    
    // Method 1: Allocate a single block for 2D array
    int* matrix1 = (int*)malloc(rows * cols * sizeof(int));
    
    if (matrix1 == NULL) {
        printf("Memory allocation failed\n");
        return;
    }
    
    // Initialize and access using row-major indexing math
    printf("Method 1 (single block):\n");
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            matrix1[i * cols + j] = i * 10 + j;
            printf("%2d ", matrix1[i * cols + j]);
        }
        printf("\n");
    }
    
    // Method 2: Array of pointers (jagged array)
    int** matrix2 = (int**)malloc(rows * sizeof(int*));
    
    if (matrix2 == NULL) {
        printf("Memory allocation failed\n");
        free(matrix1);
        return;
    }
    
    // Allocate each row
    for (int i = 0; i < rows; i++) {
        matrix2[i] = (int*)malloc(cols * sizeof(int));
        
        if (matrix2[i] == NULL) {
            // Clean up previously allocated rows
            for (int j = 0; j < i; j++) {
                free(matrix2[j]);
            }
            free(matrix2);
            free(matrix1);
            printf("Memory allocation failed for row %d\n", i);
            return;
        }
    }
    
    // Initialize and access using natural 2D indexing
    printf("\nMethod 2 (array of pointers):\n");
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            matrix2[i][j] = i * 10 + j;
            printf("%2d ", matrix2[i][j]);
        }
        printf("\n");
    }
    
    // Clean up method 2 (must free each row)
    for (int i = 0; i < rows; i++) {
        free(matrix2[i]);
    }
    free(matrix2);
    
    // Clean up method 1
    free(matrix1);
    
    matrix1 = NULL;
    matrix2 = NULL;
}

int main() {
    printf("C Memory Management Demonstration\n");
    
    demonstrate_malloc();
    demonstrate_calloc();
    demonstrate_realloc();
    demonstrate_pointer_arithmetic();
    demonstrate_memory_functions();
    demonstrate_memory_leaks();
    demonstrate_memory_pool();
    demonstrate_multidimensional_arrays();
    
    printf("\nAll demonstrations completed successfully\n");
    return 0;
}
