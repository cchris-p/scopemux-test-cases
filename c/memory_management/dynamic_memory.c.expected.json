{
  "language": "C",
  "ast": {
    "type": "translation_unit",
    "functions": [
      {
        "name": "demonstrate_malloc",
        "type": "function_definition",
        "parameters": [],
        "return_type": "void",
        "body_range": {
          "start_line": 17,
          "start_column": 0,
          "end_line": 45,
          "end_column": 1
        },
        "docstring": "Basic dynamic memory allocation with malloc"
      },
      {
        "name": "demonstrate_calloc",
        "type": "function_definition",
        "parameters": [],
        "return_type": "void",
        "body_range": {
          "start_line": 50,
          "start_column": 0,
          "end_line": 82,
          "end_column": 1
        },
        "docstring": "Using calloc for zero-initialized memory"
      },
      {
        "name": "demonstrate_realloc",
        "type": "function_definition",
        "parameters": [],
        "return_type": "void",
        "body_range": {
          "start_line": 87,
          "start_column": 0,
          "end_line": 146,
          "end_column": 1
        },
        "docstring": "Using realloc to resize allocated memory"
      },
      {
        "name": "demonstrate_pointer_arithmetic",
        "type": "function_definition",
        "parameters": [],
        "return_type": "void",
        "body_range": {
          "start_line": 151,
          "start_column": 0,
          "end_line": 193,
          "end_column": 1
        },
        "docstring": "Demonstrate pointer arithmetic and memory access"
      },
      {
        "name": "demonstrate_memory_functions",
        "type": "function_definition",
        "parameters": [],
        "return_type": "void",
        "body_range": {
          "start_line": 198,
          "start_column": 0,
          "end_line": 249,
          "end_column": 1
        },
        "docstring": "Demonstrate memory manipulation functions"
      },
      {
        "name": "demonstrate_memory_leaks",
        "type": "function_definition",
        "parameters": [],
        "return_type": "void",
        "body_range": {
          "start_line": 254,
          "start_column": 0,
          "end_line": 291,
          "end_column": 1
        },
        "docstring": "Demonstrate a common pitfall: memory leaks"
      },
      {
        "name": "pool_alloc",
        "type": "function_definition",
        "parameters": [
          {"name": "size", "type": "size_t"}
        ],
        "return_type": "void*",
        "body_range": {
          "start_line": 299,
          "start_column": 0,
          "end_line": 311,
          "end_column": 1
        },
        "docstring": "Simple memory pool allocator"
      },
      {
        "name": "pool_reset",
        "type": "function_definition",
        "parameters": [],
        "return_type": "void",
        "body_range": {
          "start_line": 313,
          "start_column": 0,
          "end_line": 315,
          "end_column": 1
        },
        "docstring": "Reset the memory pool"
      },
      {
        "name": "demonstrate_memory_pool",
        "type": "function_definition",
        "parameters": [],
        "return_type": "void",
        "body_range": {
          "start_line": 317,
          "start_column": 0,
          "end_line": 358,
          "end_column": 1
        },
        "docstring": "Demonstrate a simple custom memory pool"
      },
      {
        "name": "demonstrate_multidimensional_arrays",
        "type": "function_definition",
        "parameters": [],
        "return_type": "void",
        "body_range": {
          "start_line": 363,
          "start_column": 0,
          "end_line": 431,
          "end_column": 1
        },
        "docstring": "Demonstrate handling of multi-dimensional arrays with dynamic allocation"
      },
      {
        "name": "main",
        "type": "function_definition",
        "parameters": [],
        "return_type": "int",
        "body_range": {
          "start_line": 433,
          "start_column": 0,
          "end_line": 445,
          "end_column": 1
        }
      }
    ],
    "includes": [
      {"name": "stdio.h", "system": true, "line": 11},
      {"name": "stdlib.h", "system": true, "line": 12},
      {"name": "string.h", "system": true, "line": 13},
      {"name": "assert.h", "system": true, "line": 14}
    ],
    "docstrings": [
      {
        "type": "file_docstring",
        "text": "Demonstration of C Memory Management\n\nThis example covers:\n- Dynamic memory allocation (malloc, calloc, realloc, free)\n- Memory manipulation functions (memcpy, memmove, memset)\n- Pointer arithmetic\n- Memory safety considerations\n- Custom memory allocation patterns",
        "range": {
          "start_line": 1,
          "start_column": 0,
          "end_line": 9,
          "end_column": 3
        }
      }
    ],
    "global_variables": [
      {
        "name": "memory_pool",
        "type": "char[1024]",
        "is_static": true,
        "line": 297
      },
      {
        "name": "pool_offset",
        "type": "size_t",
        "is_static": true,
        "initial_value": "0",
        "line": 298
      }
    ],
    "macros": [
      {
        "name": "POOL_SIZE",
        "value": "1024",
        "line": 296
      }
    ],
    "memory_functions": [
      {
        "name": "malloc",
        "usage": [
          {"line": 21, "size_expression": "5 * sizeof(int)"},
          {"line": 90, "size_expression": "3 * sizeof(int)"},
          {"line": 154, "size_expression": "5 * sizeof(int)"},
          {"line": 201, "size_expression": "20"},
          {"line": 202, "size_expression": "20"},
          {"line": 255, "size_expression": "sizeof(int) * 10"},
          {"line": 272, "size_expression": "sizeof(int) * 5"},
          {"line": 367, "size_expression": "rows * cols * sizeof(int)"},
          {"line": 379, "size_expression": "rows * sizeof(int*)"},
          {"line": 392, "size_expression": "cols * sizeof(int)"}
        ]
      },
      {
        "name": "calloc",
        "usage": [
          {"line": 55, "count_expression": "5", "size_expression": "sizeof(int)"}
        ]
      },
      {
        "name": "realloc",
        "usage": [
          {"line": 107, "old_pointer": "numbers", "size_expression": "5 * sizeof(int)"},
          {"line": 126, "old_pointer": "numbers", "size_expression": "2 * sizeof(int)"}
        ]
      },
      {
        "name": "free",
        "usage": [
          {"line": 42, "pointer": "numbers"},
          {"line": 79, "pointer": "numbers"},
          {"line": 143, "pointer": "numbers"},
          {"line": 190, "pointer": "arr"},
          {"line": 245, "pointer": "buffer1"},
          {"line": 246, "pointer": "buffer2"},
          {"line": 287, "pointer": "more_data"},
          {"line": 288, "pointer": "data"},
          {"line": 394, "pointer": "matrix2[j]"},
          {"line": 397, "pointer": "matrix2"},
          {"line": 398, "pointer": "matrix1"},
          {"line": 421, "pointer": "matrix2[i]"},
          {"line": 423, "pointer": "matrix2"},
          {"line": 426, "pointer": "matrix1"}
        ]
      },
      {
        "name": "memset",
        "usage": [
          {"line": 208, "pointer": "buffer1", "value": "'A'", "size": "19"}
        ]
      },
      {
        "name": "memcpy",
        "usage": [
          {"line": 213, "destination": "buffer2", "source": "source", "size": "strlen(source) + 1"}
        ]
      },
      {
        "name": "memmove",
        "usage": [
          {"line": 219, "destination": "overlap + 1", "source": "overlap + 3", "size": "6"}
        ]
      },
      {
        "name": "memcmp",
        "usage": [
          {"line": 226, "pointer1": "str1", "pointer2": "str2", "size": "5"},
          {"line": 227, "pointer1": "str1", "pointer2": "str3", "size": "5"}
        ]
      },
      {
        "name": "pool_alloc",
        "usage": [
          {"line": 324, "size_expression": "5 * sizeof(int)"},
          {"line": 338, "size_expression": "20"}
        ]
      },
      {
        "name": "pool_reset",
        "usage": [
          {"line": 321},
          {"line": 354}
        ]
      }
    ],
    "pointer_operations": [
      {
        "operation": "pointer_arithmetic",
        "examples": [
          {"line": 166, "expression": "arr[2]", "access_type": "array_indexing"},
          {"line": 169, "expression": "*(arr + 2)", "access_type": "pointer_arithmetic"},
          {"line": 172, "expression": "*ptr", "access_type": "pointer_dereference"},
          {"line": 176, "expression": "ptr++", "access_type": "pointer_increment"},
          {"line": 180, "expression": "ptr = arr + 3", "access_type": "pointer_offset"},
          {"line": 185, "expression": "ptr2 - ptr", "access_type": "pointer_difference"}
        ]
      },
      {
        "operation": "null_checking",
        "examples": [
          {"line": 23, "expression": "if (numbers == NULL)"},
          {"line": 57, "expression": "if (numbers == NULL)"},
          {"line": 92, "expression": "if (numbers == NULL)"},
          {"line": 108, "expression": "if (new_numbers == NULL)"},
          {"line": 127, "expression": "if (new_numbers == NULL)"},
          {"line": 156, "expression": "if (arr == NULL)"},
          {"line": 204, "expression": "if (buffer1 == NULL || buffer2 == NULL)"},
          {"line": 258, "expression": "if (data == NULL)"},
          {"line": 275, "expression": "if (more_data == NULL)"},
          {"line": 300, "expression": "if (pool_offset + size > POOL_SIZE)"},
          {"line": 326, "expression": "if (numbers == NULL)"},
          {"line": 340, "expression": "if (text == NULL)"},
          {"line": 369, "expression": "if (matrix1 == NULL)"},
          {"line": 381, "expression": "if (matrix2 == NULL)"},
          {"line": 394, "expression": "if (matrix2[i] == NULL)"}
        ]
      },
      {
        "operation": "pointer_casting",
        "examples": [
          {"line": 21, "expression": "(int*)malloc(5 * sizeof(int))"},
          {"line": 55, "expression": "(int*)calloc(5, sizeof(int))"},
          {"line": 90, "expression": "(int*)malloc(3 * sizeof(int))"},
          {"line": 107, "expression": "(int*)realloc(numbers, 5 * sizeof(int))"},
          {"line": 126, "expression": "(int*)realloc(numbers, 2 * sizeof(int))"},
          {"line": 154, "expression": "(int*)malloc(5 * sizeof(int))"},
          {"line": 201, "expression": "(char*)malloc(20)"},
          {"line": 202, "expression": "(char*)malloc(20)"},
          {"line": 255, "expression": "(int*)malloc(sizeof(int) * 10)"},
          {"line": 272, "expression": "(int*)malloc(sizeof(int) * 5)"},
          {"line": 324, "expression": "(int*)pool_alloc(5 * sizeof(int))"},
          {"line": 338, "expression": "(char*)pool_alloc(20)"},
          {"line": 367, "expression": "(int*)malloc(rows * cols * sizeof(int))"},
          {"line": 379, "expression": "(int**)malloc(rows * sizeof(int*))"},
          {"line": 392, "expression": "(int*)malloc(cols * sizeof(int))"}
        ]
      },
      {
        "operation": "multidimensional_arrays",
        "examples": [
          {"line": 375, "expression": "matrix1[i * cols + j]", "access_type": "linear_indexing"},
          {"line": 376, "expression": "matrix1[i * cols + j]", "access_type": "linear_indexing"},
          {"line": 405, "expression": "matrix2[i][j]", "access_type": "2d_indexing"},
          {"line": 406, "expression": "matrix2[i][j]", "access_type": "2d_indexing"}
        ]
      }
    ],
    "memory_safety_patterns": [
      {
        "pattern": "null_check_after_allocation",
        "examples": [
          {"line": 23, "code": "if (numbers == NULL) { ... return; }"},
          {"line": 57, "code": "if (numbers == NULL) { ... return; }"},
          {"line": 92, "code": "if (numbers == NULL) { ... return; }"}
        ]
      },
      {
        "pattern": "null_pointer_after_free",
        "examples": [
          {"line": 43, "code": "numbers = NULL;"},
          {"line": 80, "code": "numbers = NULL;"},
          {"line": 144, "code": "numbers = NULL;"},
          {"line": 191, "code": "ptr = NULL;"},
          {"line": 247, "code": "buffer1 = NULL;"},
          {"line": 248, "code": "buffer2 = NULL;"},
          {"line": 289, "code": "more_data = NULL;"},
          {"line": 290, "code": "data = NULL;"},
          {"line": 428, "code": "matrix1 = NULL;"},
          {"line": 429, "code": "matrix2 = NULL;"}
        ]
      },
      {
        "pattern": "handle_realloc_failure",
        "examples": [
          {"line": 108, "code": "if (new_numbers == NULL) { ... free(numbers); return; }"},
          {"line": 127, "code": "if (new_numbers == NULL) { ... free(numbers); return; }"}
        ]
      },
      {
        "pattern": "clean_up_on_error",
        "examples": [
          {"line": 204, "code": "if (buffer1 == NULL || buffer2 == NULL) { free(buffer1); free(buffer2); return; }"},
          {"line": 275, "code": "if (more_data == NULL) { ... free(data); return; }"},
          {"line": 393, "code": "if (matrix2[i] == NULL) { for (int j = 0; j < i; j++) { free(matrix2[j]); } free(matrix2); free(matrix1); ... return; }"}
        ]
      }
    ],
    "function_calls": [
      {
        "name": "demonstrate_malloc",
        "arguments": [],
        "line": 436
      },
      {
        "name": "demonstrate_calloc",
        "arguments": [],
        "line": 437
      },
      {
        "name": "demonstrate_realloc",
        "arguments": [],
        "line": 438
      },
      {
        "name": "demonstrate_pointer_arithmetic",
        "arguments": [],
        "line": 439
      },
      {
        "name": "demonstrate_memory_functions",
        "arguments": [],
        "line": 440
      },
      {
        "name": "demonstrate_memory_leaks",
        "arguments": [],
        "line": 441
      },
      {
        "name": "demonstrate_memory_pool",
        "arguments": [],
        "line": 442
      },
      {
        "name": "demonstrate_multidimensional_arrays",
        "arguments": [],
        "line": 443
      }
    ]
  }
}
